docarray/utils/apply.py

```diff
@@ -0,0 +1,299 @@
+        you should always use `thread` backend.
+
+        .. warning::
+            When using `process` backend, your `func` should not modify elements in-place.
+            This is because the multiprocessing backend passes the variable via pickle
+            and works in another process.
+            The passed object and the original object do **not** share the same memory.
+
+    :param num_worker: the number of parallel workers. If not given, the number of
+        CPUs in the system will be used.
+    :param pool: use an existing/external process or thread pool. If given, you will
+        be responsible for closing the pool.
+    :param show_progress: show a progress bar. Defaults to False.
+
+    """
+    for i, doc in enumerate(_map(da, func, backend, num_worker, pool, show_progress)):
+        da[i] = doc
+
+
+def _map(
+    da: T,
+    func: Callable[[T_doc], T_doc],
+    backend: str = 'thread',
+    num_worker: Optional[int] = None,
+    pool: Optional[Union[Pool, ThreadPool]] = None,
+    show_progress: bool = False,
+) -> Generator[T_doc, None, None]:
+    """
+    Return an iterator that applies `func` to every Document in `da` in parallel,
+    yielding the results.
+
+    .. seealso::
+        - To return :class:`DocumentArray`, please use :meth:`.apply`.
+
+    :param da: DocumentArray to apply function to
+    :param func: a function that takes a :class:`BaseDocument` as input and outputs
+        a :class:`BaseDocument`.
+    :param backend: `thread` for multithreading and `process` for multiprocessing.
+        Defaults to `thread`.
+        In general, if `func` is IO-bound then `thread` is a good choice.
+        On the other hand, if `func` is CPU-bound, then you may use `process`.
+        In practice, you should try yourselves to figure out the best value.
+        However, if you wish to modify the elements in-place, regardless of IO/CPU-bound,
+        you should always use `thread` backend.
+
+        .. warning::
+            When using `process` backend, your `func` should not modify elements in-place.
+            This is because the multiprocessing backend passes the variable via pickle
+            and works in another process.
+            The passed object and the original object do **not** share the same memory.
+
+    :param num_worker: the number of parallel workers. If not given, the number of CPUs
+        in the system will be used.
+    :param pool: use an existing/external process or thread pool. If given, you will
+        be responsible for closing the pool.
+    :param show_progress: show a progress bar. Defaults to False.
+
+    :yield: Documents returned from `func`
+    """
+
+    if backend == 'process' and _is_lambda_or_partial_or_local_function(func):
+        raise ValueError(
+            f'Multiprocessing does not allow functions that are local, lambda or partial: {func}'
+        )
+
+    ctx_p: Union[nullcontext, Union[Pool, ThreadPool]]
```

Comment:
Let's not be stingy on chars when it comes to variable naming ;) what does this mean? context pool?